sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala

```diff
@@ -142,22 +142,22 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
 }
 
 /**
- * Replaces the expression of CurrentDatabase with the current database name.
- * Replaces the expression of CurrentCatalog with the current catalog name.
+ * Replaces the expression of CurrentDatabase, CurrentCatalog, and CurrentUser
+ * with the current values.
  */
 case class ReplaceCurrentLike(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = {
     import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
-    val currentNamespace = catalogManager.currentNamespace.quoted
-    val currentCatalog = catalogManager.currentCatalog.name()
-    val currentUser = CurrentUserContext.getCurrentUser
```

Comment:
what if these states get changed in the middle? then we don't have the snapshot semantic we previously had. Can we simply use lazy val?

Reply:
Hm I wonder, wouldn't the evaluation still be at the same time?  I thought lazy val just defers the evaluation to when its accessed, so it will be at same time as if we move the val definition inside the case?  

(I initially preferred lazy val as well, but thought these two are functionally the same).

Reply:
Got it. If the tree have multiple node with `CURRENT_LIKE`, the result may be different.
Please change them with lazy. @szehon-ho Sorry for my bad idea.

Reply:
Now we access these states for every single current_xxx function.

Reply:
Got it, thanks all for explanation, I made the update patch: https://github.com/apache/spark/pull/50028